### PR TITLE
Switch on C# pattern matching for matching types

### PR DESCRIPTION
**Pull request by hiMinju**
#25 
This PR switches to some lines using C# pattern matching.

Change like 
 `if (type == typeof(BooleanDataFrameColumn))`
to
`case Type t when t == typeof(BooleanDataFrameColumn):`

### CODE REVIEW

**Comment by dnfadmin** [comment]
[CLA assistant check](https://cla.dotnetfoundation.org/dotnet/spark?pullRequest=646) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/dotnet/spark?pullRequest=646) before we can accept your contribution.<br/><br/>:x: hiMinju [sign now](https://cla.dotnetfoundation.org/dotnet/spark?pullRequest=646)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/dotnet/spark?pullRequest=646) it.</sub>

File: src/csharp/Microsoft.Spark/Sql/ArrowArrayHelpers.cs [hunk 1]
```diff
@@ -42,57 +42,34 @@ internal static class ArrowArrayHelpers
...
+                case Type t when t == typeof(UInt64DataFrameColumn):
+                    return new UInt64DataFrameColumn("Empty");
+                case Type t when t == typeof(SingleDataFrameColumn):
+                    return new SingleDataFrameColumn("Empty");
+                case Type t when t == typeof(DoubleDataFrameColumn):
+                    return new DoubleDataFrameColumn("Empty");
+                case Type t when t == typeof(ArrowStringDataFrameColumn):
+                    return new ArrowStringDataFrameColumn("Empty");
+                default:
+                    throw new NotSupportedException($"Unknown type: {typeof(T)}");
```
Comment (imback82) [inline src/csharp/Microsoft.Spark/Sql/ArrowArrayHelpers.cs:72]: How about something like the following to simplify?
```csharp
public static DataFrameColumn CreateEmptyColumn<T>()
{
    return typeof(T) switch
    {
        Type t when t == typeof(BooleanDataFrameColumn) =>
            new BooleanDataFrameColumn("Empty"),
        Type t when t == typeof(SByteDataFrameColumn) =>
            new SByteDataFrameColumn("Empty"),
        ...
        _ => throw new NotSupportedException($"Unknown type: {typeof(T)}")
    };
}
```

File: src/csharp/Microsoft.Spark/Sql/ArrowArrayHelpers.cs [hunk 2]
```diff
@@ -113,57 +90,53 @@ private static ArrayData BuildEmptyArrayDataFromArrayType<T>()
...
+                    break;
+                case Type t when t == typeof(TimestampArray):
+                    arrowType = TimestampType.Default;
+                    break;
+                case Type t when t == typeof(StringArray):
+                    return new ArrayData(StringType.Default, 0,
+                    buffers: new[] { ArrowBuffer.Empty, ArrowBuffer.Empty, ArrowBuffer.Empty });
+                case Type t when t == typeof(BinaryArray):
+                    return new ArrayData(BinaryType.Default, 0,
+                    buffers: new[] { ArrowBuffer.Empty, ArrowBuffer.Empty, ArrowBuffer.Empty });
```
Comment (imback82) [inline src/csharp/Microsoft.Spark/Sql/ArrowArrayHelpers.cs:139]: Similarly, I think we can do the following?
```csharp
        private static ArrayData BuildEmptyArrayDataFromArrayType<T>()
        {
            return typeof(T) switch
            {
                Type t when t == typeof(BooleanArray) =>
                    BuildEmptyArrayDataFromArrowType(BooleanType.Default),
                Type t when t == typeof(Int8Array) =>
                    BuildEmptyArrayDataFromArrowType(Int8Type.Default),
                ...
                Type t when t == typeof(StringArray) =>
                    BuildEmptyArrayDataFromArrowType(StringType.Default),
                ...
                _ => throw new NotSupportedException($"Unknown type:  {typeof(T)}")
            };
        }
```

Comment (hiMinju) [reply src/csharp/Microsoft.Spark/Sql/ArrowArrayHelpers.cs:139]: I think it looks better!
I'll fix and commit again.

Thanks @imback82

Comment (hiMinju) [reply src/csharp/Microsoft.Spark/Sql/ArrowArrayHelpers.cs]: I reflected your reviews!

File: src/csharp/Microsoft.Spark/Sql/ArrowArrayHelpers.cs [hunk 3]
```diff
@@ -41,59 +41,33 @@ internal static class ArrowArrayHelpers
...
+                    new BooleanDataFrameColumn("Empty"),
+                Type t when t == typeof(SByteDataFrameColumn) =>
+                    new SByteDataFrameColumn("Empty"),
+                Type t when t == typeof(ByteDataFrameColumn) =>
+                    new ByteDataFrameColumn("Empty"),
+                Type t when t == typeof(Int16DataFrameColumn) =>
+                    new Int16DataFrameColumn("Empty"),
+                Type t when t == typeof(UInt16DataFrameColumn) =>
+                    new UInt16DataFrameColumn("Empty"),
+                Type t when t == typeof(Int32DataFrameColumn) =>
```
Comment (imback82) [inline src/csharp/Microsoft.Spark/Sql/ArrowArrayHelpers.cs:56]: missing uint32?

Comment (hiMinju) [reply src/csharp/Microsoft.Spark/Sql/ArrowArrayHelpers.cs:56]: Sorry.. I missed it.
I just add it!
 Thanks for letting me know.

**Review by imback82 — APPROVED** [review]
LGTM, thanks @hiMinju!

File: src/csharp/Microsoft.Spark/Sql/ArrowArrayHelpers.cs [hunk 4]
```diff
@@ -41,59 +41,35 @@ internal static class ArrowArrayHelpers
...
-            else
-            {
-                throw new NotSupportedException($"Unknown type: {typeof(T)}");
-            }
+            return typeof(T) switch
+            {
+                Type t when t == typeof(BooleanDataFrameColumn) =>
+                    new BooleanDataFrameColumn("Empty"),
+                Type t when t == typeof(SByteDataFrameColumn) =>
+                    new SByteDataFrameColumn("Empty"),
```
Comment (suhsteve) [inline src/csharp/Microsoft.Spark/Sql/ArrowArrayHelpers.cs:49]: Would it look nicer if we moved the type var outside the switch and used _ inplace of them ?
```csharp
            Type type = typeof(T);
            return type switch
            {
                _ when type == typeof(BooleanDataFrameColumn) =>
                    new BooleanDataFrameColumn("Empty"),
                _ when type == typeof(SByteDataFrameColumn) =>
                    new SByteDataFrameColumn("Empty"),
...
```

Comment (hiMinju) [reply src/csharp/Microsoft.Spark/Sql/ArrowArrayHelpers.cs]: I think so too. It looks nicer.
I just changed it!

Thanks @suhsteve

File: src/csharp/Microsoft.Spark/Sql/ArrowArrayHelpers.cs [hunk 5]
```diff
@@ -42,58 +42,35 @@ internal static class ArrowArrayHelpers
...
+                _ when type == typeof(UInt64DataFrameColumn) =>
+                    new UInt64DataFrameColumn("Empty"),
+                _ when type == typeof(SingleDataFrameColumn) =>
+                    new SingleDataFrameColumn("Empty"),
+                _ when type == typeof(DoubleDataFrameColumn) =>
+                    new DoubleDataFrameColumn("Empty"),
+                _ when type == typeof(ArrowStringDataFrameColumn) =>
+                    new ArrowStringDataFrameColumn("Empty"),
+
+                _ => throw new NotSupportedException($"Unknown type: {typeof(T)}")
```
Comment (suhsteve) [inline src/csharp/Microsoft.Spark/Sql/ArrowArrayHelpers.cs:72]:
```suggestion
                _ => throw new NotSupportedException($"Unknown type: {type}")
```

**Review by suhsteve — APPROVED** [review]
LGTM.  Thanks @hiMinju !

**Review by imback82 — APPROVED** [review]
Thanks all! Merging to master

**Pull request merged**